docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,388 @@
+.. _numpy_compatibility:
+
+.. testsetup:: *
+
+   import numpy as np
+   import torch
+
+NumPy Compatibility
+===================
+
+.. note::
+
+   **TL;DR:**
+
+   - Be explicit about obtaining and using NumPy arrays
+   - `This tutorial`_ shows a more concrete example of interoperability between
+     NumPy and PyTorch
+
+At the onset it should be stressed that PyTorch tensors and NumPy arrays are
+fundamentally different types. However, in the interests of being accessible to
+the wider scientific python ecosystem, there are implicit and explicit
+conversion methods. The major considerations to keep in mind are:
+
+- **Where** is the structure allocated?
+    * This may be on a compute device (e.g. a GPU) or the CPU
+- **What** is the underlying data?
+    * In particular, several functions will either return a copy or a view
+- **How** are modifications to the data handled?
+    * The data returned may be read-only, or may be writable
+
+From the end-user perspective, it is most desirable to have zero copy methods,
+however, for exploratory analyses and their visualizations in particular, it is
+often more important to have guaranteed conversions, even if copies are created.
+
+Basic interoperability
+----------------------
+
+NumPy arrays have no conception of the GPU, and so all conversions require a
+``cpu`` tensor. Similarly, the **computational graph** cannot be represented in
+```numpy``, and so the pre-requisite to have an operation or object without one,
+i.e. only leaf nodes are supported.
+
+.. doctest::
+
+   >>> b_torch = torch.ones(3)
+   >>> assert(b_torch.requires_grad == False)
+   >>> assert(b_torch.is_cuda == False)
+   >>> b_torch.numpy()
+   array([1., 1., 1.], dtype=float32)
+
+PyTorch will raise exceptions instead of allowing operations that may lead to
+incorrect results. As the conversion to a NumPy array may potentially be a lossy
+operation (the computational graph), there are no inbuilt operations to do
+something which unconditionally returns a NumPy array. Practically speaking, a
+function can be created locally to return a NumPy array, for say, visualizations.
+
+.. code-block:: python
+
+   # Possibly lossy function for visualizations
+   def uncond_numpy(torch_tensor):
+        if torch_tensor.is_cuda:
+           torch_tensor = torch_tensor.cpu()
+        if torch_tensor.requires_grad == False:
+            return torch_tensor.numpy()
+        else:
+            return torch_tensor.detach().numpy()
+
+Operations
+----------
+
+All ``torch`` operators will helpfully fail with a ``TypeError`` if called with
```

Comment:
I'm not sure this is true

Reply:
It certainly seems to be so:
```python
>>> b_torch = torch.ones(3)
>>> a_np = np.ones(3)
>>> torch.cross(b_torch, a_np)
>>> torch.add(b_torch, a_np)
```
Both of these will raise `argument 'other' (position 2) must be Tensor, not numpy.ndarray`

Reply:
I've tried to reword this but I'm not sure it is as clear as possible.